docs/readmes/bazel/bazel.md

```diff
@@ -0,0 +1,510 @@
+
+Bazel rules that are used in the Magma AGW are:
+
+- `cc_library`, `py_library` and `go_library` are used to group C, C++, Python and Go code respectively.
+- `cc_binary`, `py_binary`, `go_binary` are used for the definitions of the C, C++, Python and Go AGW services and scripts, e.g. the `//lte/gateway/c/core:agw_of` target for MME, the `//lte/gateway/python/magma/pipelined` target for pipelined or the `//feg/gateway/services/envoy_controller` target for the envoy controller.
+- `cc_test` and `go_test` are used for C, C++ and Go unit tests.
+- `pytest_test` is a wrapper around `py_test`. This rule type is used for Python unit tests, the [LTE integration tests](#lte-integration-tests) and the [Python sudo tests](#python-sudo-tests).
+- `proto_library`, `cpp_proto_library`, `python_proto_library`, `cpp_grpc_library`, `python_grpc_library` are used to group proto code, as well as for providing proto and grpc code for C++ and Python respectively.
+
+#### Target visibility
+
+Bazel targets have a [visibility attribute](https://bazel.build/concepts/visibility). Per default a target's visibility is set to private and the target is only accessible for targets within the same package. Possible values include (among others) a public visibility `//visibility:public` (accessible from all packages) or package specific visibility `//foo/bar:__pkg__`, to grant targets in package `//foo/bar` access. The visibility can be set explicitly for a target or globally for the whole package.
+
+The following coding guidelines should be applied when adding new targets. One important motivation for the guidelines below is based on encapsulation, i.e. keeping a clean architecture by restricting access to targets. For example, service libraries should only be accessed by the service itself. Proto and utility libraries can be accessed by any service (with certain restrictions if appropriate - e.g. the `all_python_proto` target should only be accessed by the `state` service).
+
+- Protos `BUILD.bazel` files: The global package visibility is set to public as compiled protos libraries need to be accessed from a lot of packages. The visibility of native proto libraries is set to private or is restricted to package visibility for other protos packages.
+- Service `BUILD.bazel` files: The global package visibility is set to private (i.e. default). The binaries should explicitly be set to private for readability reasons, even if this means redundancy. Libraries should explicitly be set to private or package visibility (e.g. for testing) when needed. The public visibility should be used only if necessary.
+- Utility Packages: In the `orc8r/configuration` and `orc8r/common` `BUILD.bazel` files the global package visibility is set to public.
+
+### Target command line syntax
+
+A complete description of the [syntax for specifying targets](https://bazel.build/run/build) can be found in the official bazel documentation. Here we provide a list of some useful examples:
```

Comment:
"syntax for specifying targets" -> "command line syntax for building targets"?

Reply:
I would prefer to keep `specifying` as this part also applies to specifying targets for testing, coverage etc. In later sections we refer to this bit for how to specify targets.

Reply:
Imho a target is a build description using a rule, i.e., the syntax of a target describes a construct
```
my_rule(
  name = foo,
  attribute1 = [...],
  ...,
)
```
The target can be build by using a bazel verb on the label of the target. -> Do you mean `syntax for specifying labels`?

Reply:
Hm perhaps a better phrasing is `syntax for specifying target patterns`, i.e.
![Screenshot from 2023-01-09 09-54-03](https://user-images.githubusercontent.com/34488763/211270823-9a36a7ce-932c-493c-8247-4e4b99b65b68.png)

Reply:
Hmm the term "pattern" seems to be used because of `...` and `*`. But: this indicates that Bazel is using the term "target" synonym for a target in a BUILD.bazel file and the label that is called via command line - would not be my choice, but then your initial phrasing is imho OK.